horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/RunServiceImpl.java

```diff
@@ -1118,14 +1122,38 @@ public void recalculateAll(String fromStr, String toStr) {
         }
     }
 
+    /**
+     * Transforms the data for a given run by applying applicable schemas and transformers.
+     * It ensures any existing datasets for the run are removed before creating new ones,
+     * handles timeouts for ongoing transformations, and creates datasets with the transformed data.
+     *
+     * @param runId the ID of the run to transform
+     * @param isRecalculation flag indicating if this is a recalculation
+     * @return the number of datasets created, or 0 if the run is invalid or not found or already ongoing
+     */
     @WithRoles(extras = Roles.HORREUM_SYSTEM)
     @Transactional
     int transform(int runId, boolean isRecalculation) {
         if (runId < 1) {
             log.errorf("Transformation parameters error: run %s", runId);
             return 0;
         }
+
         log.debugf("Transforming run ID %d, recalculation? %s", runId, Boolean.toString(isRecalculation));
+        int numDatasets = 0;
+
+        // check whether there is an ongoing transformation on the same runId
+        TestService.RecalculationStatus status = new TestService.RecalculationStatus(1);
+        TestService.RecalculationStatus prev = transformations.putIfAbsent(runId, status);
+        // ensure the transformation is removed, with this approach we should be sure
+        // it gets removed even if transaction-level exception occurs, e.g., timeout
+        Util.registerTxSynchronization(tm, txStatus -> transformations.remove(runId, status));
```

Comment:
I couldn't use `remove(runId, null)` as the `remove` methods expect a non-null value to properly remove the object:

```java
   public boolean remove(Object key, Object value) {
        if (key == null)
            throw new NullPointerException();
        return value != null && replaceNode(key, null, value) != null;
    }
```